### PR TITLE
Accessory - Add conditional blocking of switchable attachments

### PR DESCRIPTION
**Pull request by PabstMirror**
Allows conditional code to block some attachment modes
e.g. some IR illuminators can be physically locked out of high power modes

### CODE REVIEW

File: addons/accessory/fnc_addAttachementCondition.sqf [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#include "script_component.hpp"
+/* ----------------------------------------------------------------------------
+Function: CBA_fnc_addAttachementCondition
+
+Description:
+    Adds condition to be able to swtich to a attachement.
```
Comment (rautamiekka) [inline addons/accessory/fnc_addAttachementCondition.sqf:6]:
```suggestion
    Adds condition to be able to switch to an attachment.
```

File: addons/accessory/fnc_addAttachementCondition.sqf [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+Author:
+    PabstMirror
+---------------------------------------------------------------------------- */
+
+params [["_item", "", [""]], ["_condition", {true}, [{}]]];
+
+if (!isClass (configfile >> "CfgWeapons" >> _item)) exitWith { ERROR_1("Item not found [%1]", _item); };
+
+private _useageArray = GVAR(usageHash) getOrDefault [_item, [], true];
+_useageArray pushBack _condition;
```
Comment (rautamiekka) [inline addons/accessory/fnc_addAttachementCondition.sqf:29]:
```suggestion
private _usageArray = GVAR(usageHash) getOrDefault [_item, [], true];
_usageArray pushBack _condition;
```

File: addons/accessory/fnc_switchAttachment.sqf [hunk 3]
```diff
@@ -41,17 +41,28 @@ private _currWeaponType = call {
...
-};
-if (_configs isNotEqualTo []) then {
-    _switchItem = getText (_configs select 0);
+private _testItem = _currItem;
+while {_testItem != ""} do {
+    // Get the next/previous item from the attachement's config, but ignore inherited values
+    private _configs = if (_switchTo == "next") then {
+        configProperties [_cfgWeapons >> _testItem, "configName _x == 'MRT_SwitchItemNextClass'", false];
+    } else {
+        configProperties [_cfgWeapons >> _testItem, "configName _x == 'MRT_SwitchItemPrevClass'", false];
```
Comment (rautamiekka) [inline addons/accessory/fnc_switchAttachment.sqf:52]:
```suggestion
    private _configs = if (_switchTo isEqualTo "next") then {
        configProperties [_cfgWeapons >> _testItem, "configName _x isEqualTo 'MRT_SwitchItemNextClass'", false];
    } else {
        configProperties [_cfgWeapons >> _testItem, "configName _x isEqualTo 'MRT_SwitchItemPrevClass'", false];
```

File: addons/accessory/fnc_switchAttachment.sqf [hunk 4]
```diff
@@ -82,8 +93,8 @@ if (!isNil "_switchItem") then {
             }, [_unit, _currItem, _switchItem, _currWeaponType]] call CBA_fnc_execNextFrame;
         };
     };
-    private _switchItemHintText = getText (__cfgWeapons >> _switchItem >> "MRT_SwitchItemHintText");
-    private _switchItemHintImage = getText (__cfgWeapons >> _switchItem >> "picture");
+    private _switchItemHintText = getText (_cfgWeapons >> _switchItem >> "MRT_SwitchItemHintText");
+    private _switchItemHintImage = getText (_cfgWeapons >> _switchItem >> "picture");
```
Comment (rautamiekka) [inline addons/accessory/fnc_switchAttachment.sqf:97]:
```suggestion
    private _configSwitchItem = _cfgWeapons >> _switchItem;
    private _switchItemHintText = getText (_configSwitchItem >> "MRT_SwitchItemHintText");
    private _switchItemHintImage = getText (_configSwitchItem >> "picture");
```

File: addons/accessory/fnc_switchableAttachments.sqf [hunk 5]
```diff
@@ -39,4 +39,10 @@ while {
 } do {};
 
 _forward = _forward + _backward;
-_forward arrayIntersect _forward // return
+_forward = _forward arrayIntersect _forward;
+_forward select {
+    private _item = _x;
+    private _useageArray = GVAR(usageHash) getOrDefault [_x, []];
+    (_useageArray findIf {([_item] call _x) isEqualTo false}) == -1 // none returned false
```
Comment (rautamiekka) [inline addons/accessory/fnc_switchableAttachments.sqf:46]:
```suggestion
    private _usageArray = GVAR(usageHash) getOrDefault [_x, []];
    (_usageArray findIf {([_item] call _x) isEqualTo false}) == -1 // none returned false
```

File: addons/accessory/fnc_switchAttachment.sqf [hunk 6]
```diff
@@ -41,17 +41,28 @@ private _currWeaponType = call {
...
+    };
+    
+    if (_configs isEqualTo []) then {
+        _testItem = "";
+    } else {
+        _testItem = getText (_configs select 0);
+        if (_testItem == "") exitWith {};
+        if (_testItem == _currItem) exitWith { _testItem = ""; }; // same as start (full loop)
+        private _useageArray = GVAR(usageHash) getOrDefault [_testItem, []];
+        if ((_useageArray findIf {([_testItem] call _x) isEqualTo false}) == -1) then { // none returned false
```
Comment (rautamiekka) [inline addons/accessory/fnc_switchAttachment.sqf:62]:
```suggestion
        if (_testItem isEqualTo _currItem) exitWith { _testItem = ""; }; // same as start (full loop)
        private _usageArray = GVAR(usageHash) getOrDefault [_testItem, []];
        if ((_usageArray findIf {([_testItem] call _x) isEqualTo false}) == -1) then { // none returned false
```

**Review by rautamiekka — CHANGES_REQUESTED** [review]

File: addons/accessory/fnc_switchAttachment.sqf (hunk 3 above)
Comment (PabstMirror) [reply addons/accessory/fnc_switchAttachment.sqf:52]: making this case sensitive could break things

File: addons/accessory/fnc_switchAttachment.sqf (hunk 6 above)
Comment (PabstMirror) [reply addons/accessory/fnc_switchAttachment.sqf:62]: making this case sensitive could break things

**Review by jonpas — APPROVED** [review]

File: addons/accessory/CfgFunctions.hpp [hunk 7]
```diff
@@ -1,6 +1,7 @@
 class CfgFunctions {
     class CBA {
         class Inventory {
+            PATHTO_FNC(addAttachementCondition);
```
Comment (jonpas) [inline addons/accessory/CfgFunctions.hpp:4]:
```suggestion
            PATHTO_FNC(addAttachmentCondition);
```

File: addons/accessory/fnc_addAttachementCondition.sqf [hunk 8]
```diff
@@ -0,0 +1,29 @@
+#include "script_component.hpp"
+/* ----------------------------------------------------------------------------
+Function: CBA_fnc_addAttachementCondition
```
Comment (jonpas) [inline addons/accessory/fnc_addAttachementCondition.sqf:3]:
```suggestion
Function: CBA_fnc_addAttachmentCondition
```

File: addons/accessory/fnc_addAttachementCondition.sqf [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+Parameters:
+    0: _item - Attachment classname <STRING>
+    1: _condition - Code (return false if not allowed) <CODE>
+
+Returns:
+    None
+
+Examples:
+    (begin example)
+        ["ACE_acc_pointer_red", { false }] call CBA_fnc_addAttachementCondition
```
Comment (jonpas) [inline addons/accessory/fnc_addAttachementCondition.sqf:17]:
```suggestion
        ["ACE_acc_pointer_red", { false }] call CBA_fnc_addAttachmentCondition
```

File: addons/accessory/fnc_switchAttachment.sqf [hunk 10]
```diff
@@ -41,17 +41,28 @@ private _currWeaponType = call {
...
-private _configs = if (_switchTo == "next") then {
-    configProperties [__currItem, "configName _x == 'MRT_SwitchItemNextClass'", false];
-} else {
-    configProperties [__currItem, "configName _x == 'MRT_SwitchItemPrevClass'", false];
-};
-if (_configs isNotEqualTo []) then {
-    _switchItem = getText (_configs select 0);
+private _testItem = _currItem;
+while {_testItem != ""} do {
+    // Get the next/previous item from the attachement's config, but ignore inherited values
```
Comment (jonpas) [inline addons/accessory/fnc_switchAttachment.sqf:48]:
```suggestion
    // Get the next/previous item from the attachment's config, but ignore inherited values
```

**Review by jonpas — CHANGES_REQUESTED** [review]

**Review by jonpas — APPROVED** [review]

**Pull request merged**